ADAPT - converted to es6 and did some tidy up around show correct answer

eleanor-heath opened the pull request:
Contributing towards OS code: adaptlearning/adapt_framework#2808
"Convert all core components to ES6 classes #2808"

Converted to es6 and did some tidy up around show correct answer.

cahirodoherty-learningpool reviewed (approved):
👀

oliverfoster commented on js/sliderView.js:
```diff
@@ -180,107 +182,107 @@ define([
+    onCannotSubmit() { }
 
-    setSliderValue: function (value) {
+    setSliderValue(value) {
```
This function is an example of an easy "return early".
The main body of this function is:
```js
this.$slider.val(value).change();
```

The main body should only be executed if the given condition is met:
```js
if (this.$slider) { }
```
If we therefore invert the condition and return early, we can reduce the nesting and reduce number of lines whilst increasing the clarity:
```js
if (!this.$slider) return;
this.$slider.val(value).change();
```

oliverfoster commented on js/sliderView.js:
```diff
@@ -289,74 +291,77 @@ define([
+      const answers = [];
+      this.showScaleMarker(false);
 
+      //are we dealing with a single correct answer or a range?
```
Double existence comparison can be simplified:
```js
      const correctAnswer = this.model.get('_correctAnswer');
      if (correctAnswer !== undefined) {
          answers.push(correctAnswer);
      }
```

ethan-lp reviewed (approved):
👀

oliverfoster commented on js/sliderView.js:
```diff
@@ -289,74 +291,77 @@ define([
-        while (answer <= topOfRange) {
-          answers.push(answer);
-          answer += step;
+      if (answers.length === 0) {
```
Use variables to increase clarity and reduce nesting. Apply "return early" principle.
```js
      const correctRange = this.model.has('_correctRange');
      const hasCorrectRange = (correctRange !== undefined);

      if (!hasCorrectAnswer && !hasCorrectRange) {
        console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
      }
      
      if (!hasCorrectAnswer && hasCorrectRange) {
        const bottom = correctRange ._bottom;
        const top =correctRange ._top;
        const step = (this.model.has('_scaleStep') ? this.model.get('_scaleStep') : 1);

        if (bottom !== undefined && top !== undefined) {
          const answer = this.model.get('_correctRange')._bottom;
          const topOfRange = this.model.get('_correctRange')._top;
          while (answer <= topOfRange) {
            answers.push(answer);
            answer += step;
          }
        }
      }
```

eleanor-heath replied:
all great suggestions thanks, addressing them currently.

would you oppose if I was to make `if (!hasCorrectAnswer && hasCorrectRange) {`
`else if (hasCorrectRange) {` ?

oliverfoster replied:
Nope I wouldn't oppose, that seems fine.

tomgreenfield replied:
Thanks @oliverfoster for the suggestion – the more I read the code, the more I feel there's enough going on to warrant a separate function. Something like:
```js
getCorrectAnswers() {
  const answerSingle = this.model.get('_correctAnswer');
  if (answerSingle) {
    return [ answerSingle ];
  }
  const answerMultiple = this.model.get('_correctRange');
  const answers = [];
  if (!answerMultiple) {
    console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
    return answers;
  }
  const bottom = answerMultiple._bottom;
  const top = answerMultiple._top;
  if (bottom === undefined || top === undefined) {
    return answers;
  }
  let answer = bottom;
  const step = this.model.get('_scaleStep') || 1;
  while (answer <= top) {
    answers.push(answer);
    answer += step;
  }
  return answers;
}
```

oliverfoster replied:
Yea, that block in the middle does seem somewhat out of place.

eleanor-heath replied:
I like this idea - splits the code up a bit meaning you have a show and a get function 👍 

I was also thinking that we should only attempt to animate to position/show model answer and set slider value when there are answers. This means we could return after logging out the warning when there are no answers.

For example, 

`const answers = this.getCorrectAnswers();
  if (!answers.length) {
    console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
    return
  } else {
    const middleAnswer = answers[Math.floor(answers.length / 2)];
    this.animateToPosition(this.mapIndexToPixels(this.getIndexFromValue(middleAnswer)));
    this.showModelAnswers(answers);
    this.setSliderValue(middleAnswer);
  }
` 

Let me know your thoughts!

oliverfoster commented on js/sliderView.js:
```diff
@@ -365,26 +370,26 @@ define([
       if (this.model.get('_showScaleIndicator')) {
         this.showNumber(show);
-        if(show) {
+        if (show) {
```
```js
$scaleMarker.toggleClass('display-block', show);
```

oliverfoster commented on js/sliderView.js:
```diff
@@ -180,107 +182,107 @@ define([
 
     //Labels are enabled in slider.hbs. Here we manage their containing div.
-    showLabels: function () {
+    showLabels() {
```
Return early.

tomgreenfield commented on js/sliderView.js:
```diff
@@ -289,74 +291,77 @@ define([
-    selectItem: function(itemIndex, noFocus) {
-      _.each(this.model.get('_items'), function(item, index) {
+    selectItem(itemIndex) {
+      this.model.get('_items').forEach((item, index) => {
```
Could refactor to remove the loop:
```js
const item = this.model.get('_items')[itemIndex];
if (!item) return;
item.selected = true;
this.model.set('_selectedItem', item);
this.$('.js-slider-item-input').attr({
  'value': item.value,
  'aria-valuenow': item.value
});
```

eleanor-heath replied:
I have resolved all of your comments. Just a note on this, the first item is defaulted to be selected so have called on deselectAllItems before I select the chosen item to be sure there is not anything currently selected. If we dont do this the value of isCorrect could be wrong, so just to be sure I want to deselect them all and not just the first item.

tomgreenfield reviewed (commented):
:eye:

eleanor-heath commented:
@tomgreenfield @oliverfoster I have addressed all of your comments, can you take another look please when you have the chance.

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,78 @@ define([
+    
+    isCorrect() {
+      let numberOfCorrectAnswers = 0;
+      this.get('_items').forEach(item => {
```
```suggestion
      this.set('_isAtLeastOneCorrectSelection', false);
      this.get('_items').forEach(item => {
```

oliverfoster replied:
Part 1 to remove the weird `return this.get('_isAtLeastOneCorrectSelection') ? true : false;` ternary

tomgreenfield replied:
Using a filter would make things a lot simpler:
```js
isCorrect() {
  const numberOfCorrectAnswers = this.get('_items').filter(({ selected, correct }) => selected && correct).length;

  this.set('_isAtLeastOneCorrectSelection', numberOfCorrectAnswers > 0);
  this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);

  return this.get('_isAtLeastOneCorrectSelection');
}
```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,78 @@ define([
 
       this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
 
       return this.get('_isAtLeastOneCorrectSelection') ? true : false;
```
```suggestion
      return this.get('_isAtLeastOneCorrectSelection');
```

oliverfoster replied:
Part 2

oliverfoster commented on js/sliderView.js:
```diff
@@ -180,107 +177,105 @@ define([
-          this.model.get('_showScaleIndicator') ? {visibility: 'hidden'} : {display: 'none'}
+        $markers.eq(0).addClass('u-display-none');
+        this.$('.js-slider-number').addClass(
+          this.model.get('_showScaleIndicator') ? 'visibility-hidden' : 'u-display-none'
```
```suggestion
          this.model.get('_showScaleIndicator') ? 'u-visibility-hidden' : 'u-display-none'
```

oliverfoster replied:
[u-visibility-hidden](https://github.com/adaptlearning/adapt_framework/blob/master/src/core/less/_defaults/utility.less#L32)

oliverfoster commented on js/sliderView.js:
```diff
@@ -180,107 +177,105 @@ define([
-      this.$('.slider-handle').css({left: left + 'px'});
-      this.$('.js-slider-number-selection').css({left: left + 'px'});
+    showLabels() {
+      if (this.model.get('labelStart') && this.model.get('labelEnd')) return;
```
```suggestion
      if (this.model.get('labelStart') || this.model.get('labelEnd')) return;
```
You need to invert the operators as well

oliverfoster commented on js/sliderView.js:
```diff
@@ -289,102 +284,101 @@ define([
 
-        var $element = $(this.$('.js-slider-model-range .slider__number-model-answer')[index]);
-        var startingLeft = this.mapIndexToPixels(this.getIndexFromValue(this.model.get('_selectedItem').value));
+        const $element = $(this.$('.js-slider-model-range .slider__number-model-answer')[index]);
```
```suggestion
        const $element = this.$('.js-slider-model-range .slider__number-model-answer').eq(index);
```

oliverfoster replied:
[jQuery.eq](https://api.jquery.com/eq/)

cahirodoherty-learningpool reviewed (approved):
👀

eleanor-heath commented:
@oliverfoster can you merge this please (I don't have the permissions to)

oliverfoster commented:
@tomgreenfield needs to finish his review off with an approval

tomgreenfield commented on js/sliderView.js:
```diff
@@ -3,94 +3,91 @@ define([
-    },
+      this.$('.slider__widget').toggleClass('is-disabled', !isEnabled);
+      this.$slider.prop('disabled', !isEnabled).rangeslider('update', true);
+      return;
```
Please remove the `return`.

tomgreenfield commented on js/sliderView.js:
```diff
@@ -289,102 +284,101 @@ define([
-      var $parentDiv = this.$('.js-slider-model-range');
-      _.each(correctAnswerArray, function(correctAnswer, index) {
+    getCorrectAnswers() {
+      //are we dealing with a single correct answer or a range?
```
Please add space before comment (linter will complain).

eleanor-heath commented:
Hi @tomgreenfield, I've addressed all of your comments now. Now that I've managed to update to ES6 classes as well as do some general tidy-up, I think further refactors should be open in another tech-debt ticket.

Can you please re-review when you get a chance?

oliverfoster commented on js/sliderView.js:
```diff
@@ -394,7 +388,7 @@ define([
     }
 
-  });
+  };
```
Can you remove the `;` from the end of both classes? They're not necessary after a code block.
```js
if () {
}

class Name {
}

for () {
}
```

oliverfoster replied:
```suggestion
  }
```

oliverfoster commented on js/sliderView.js:
```diff
@@ -180,107 +176,105 @@ define([
-    },
+    }
 
     //Labels are enabled in slider.hbs. Here we manage their containing div.
```
```suggestion
    // Labels are enabled in slider.hbs. Here we manage their containing div.
```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
     }
 
-  });
+  };
```
```suggestion
  }
```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -2,38 +2,37 @@ define([
+      const start = this.get('_scaleStart');
+      const end = this.get('_scaleEnd');
+      const step = this.get('_scaleStep') || 1;
+      
```
```suggestion

```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
-    },
+    }
 
     //This preserves the state of the users answers for returning or showing the users answer
```
```suggestion
    // This preserves the state of the users answers for returning or showing the users answer
```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
+    }
+    
```
```suggestion

```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
+    
+    isCorrect() {
+      const numberOfCorrectAnswers = this.get('_items').filter(({ selected, correct }) => selected && correct).length;
+    
```
```suggestion

```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
+    
+      this.set('_isAtLeastOneCorrectSelection', numberOfCorrectAnswers > 0);
       this.set('_numberOfCorrectAnswers', numberOfCorrectAnswers);
+    
```
```suggestion

```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -2,38 +2,37 @@ define([
   'core/js/adapt',
   'core/js/models/questionModel'
 ], function(Adapt, QuestionModel) {
+  class SliderModel extends QuestionModel {
```
```suggestion

  class SliderModel extends QuestionModel {
```

oliverfoster reviewed (changes requested):
Linter (ESLint plugin in VSCode) has some warnings.

If you're using VSCode it might be worth installing the ESLint plugin and adding these settings:
```json
{
  "editor.trimAutoWhitespace": true,
  "editor.renderWhitespace": "all",
  "files.trimTrailingWhitespace": true,
  "files.insertFinalNewline": true,
  "[markdown]": {
    "files.trimTrailingWhitespace": false
  }
}
```

oliverfoster commented on js/sliderModel.js:
```diff
@@ -70,84 +69,73 @@ define([
+    }
 
+    deselectAllItems() {
+      this.get('_items').forEach(item => item.selected = false);
```
Direct assignment returns aren't allowed
```suggestion
      this.get('_items').forEach(item => (item.selected = false));
```

oliverfoster commented:
I reckon `getCorrectAnswers` should be a model function rather than a view function?

Change call to `this.model.getCorrectAnswers();`
```js
    showCorrectAnswer() {
      this.showScaleMarker(false);

      const answers = this.model.getCorrectAnswers();
      if (answers.length === 0) {
        console.log('adapt-contrib-slider::WARNING: no correct answer or correct range set in JSON');
        return;
      }

      const middleAnswer = answers[Math.floor(answers.length / 2)];
      this.animateToPosition(this.mapIndexToPixels(this.getIndexFromValue(middleAnswer)));
      this.showModelAnswers(answers);
      this.setSliderValue(middleAnswer);
    }
```

Move to model and change `this.model.` to `this.`
```js

    getCorrectAnswers() {

      // are we dealing with a single correct answer or a range?
      const answerSingle = this.get('_correctAnswer');
      const answers = [];
      if (answerSingle) {
        return [ answerSingle ];
      }
      const answerMultiple = this.get('_correctRange');
      if (!answerMultiple) {
        return answers;
      }
      const bottom = answerMultiple._bottom;
      const top = answerMultiple._top;
      if (bottom === undefined || top === undefined) {
        return answers;
      }
      let answer = bottom;
      const step = this.get('_scaleStep') || 1;
      while (answer <= top) {
        answers.push(answer);
        answer += step;
      }
      return answers;
    }
```

eleanor-heath commented:
Hi @oliverfoster, i agree that these are good suggestions but would argue that they too are out of scope of this PR.

The depth of my changes were to introduce ES6 standards. Everything else that I have addressed was a quick tidy-up, but most of your suggestions are on code that is not necessary to change at this point.
If you would prefer, I can undo my tidy-up and only include the ES6 changes so that you can open a new PR to address any concerns that you may have.

oliverfoster commented:
@eleanor-heath The linting problems break the project rules so you'll need to fix those. I have suggested above that you can modify your editor to help with those in future. Moving your new `getCorrectAnswers` function to the model won't take a minute to do and it is a model function.

The other stuff about items, indexes and `ItemComponentModel` is definitely out of scope, I agree.

moloko commented on bower.json:
```diff
@@ -1,7 +1,7 @@
   "name": "adapt-contrib-slider",
-  "version": "3.0.0",
-  "framework": ">=5",
+  "version": "3.0.1",
```
As this is more than just a 'bug fix'...
```suggestion
  "version": "3.1.0",
```

tomgreenfield reviewed (approved):
:eye:
